parachain/primitives/router/src/inbound/mod.rs

```diff
@@ -70,135 +68,172 @@ pub enum Destination {
 	ForeignAccountId20 { para_id: u32, id: [u8; 20] },
 }
 
-impl From<MessageV1> for Xcm<()> {
-	fn from(val: MessageV1) -> Self {
-		val.message.convert(val.chain_id)
+pub struct VersionedMessageToXcmConverter<G, F> {
+	_phantom: PhantomData<(G, F)>,
+}
+
+pub trait XcmFeeGetter {
+	// Measures the maximum amount of gas a command will require
+	fn fee(command: &Command) -> u128;
+}
+
+/// A meter that assigns a constant amount of gas for the execution of a command
+pub struct ConstantFeeForInboundMessage;
+
+impl XcmFeeGetter for ConstantFeeForInboundMessage {
+	fn fee(command: &Command) -> u128 {
+		match command {
+			Command::RegisterToken { .. } => 2_000_000_000,
+			Command::SendToken { .. } => 1_000_000_000,
+		}
 	}
 }
 
-impl Command {
-	pub fn convert(self, chain_id: u64) -> Xcm<()> {
-		let network = NetworkId::Ethereum { chain_id };
-		// TODO (SNO-582): The fees need to be made configurable and must match the weight
-		// required by the generated XCM script when executed on the foreign chain.
-		let buy_execution_fee_amount = 2_000_000_000;
-		let buy_execution_fee = MultiAsset {
-			id: Concrete(MultiLocation::parent()),
-			fun: Fungible(buy_execution_fee_amount),
-		};
+impl XcmFeeGetter for () {
+	fn fee(_: &Command) -> u128 {
+		0
+	}
+}
 
-		let create_instructions = |origin_location: Junction| -> Vec<Instruction<()>> {
-			vec![
-				UniversalOrigin(GlobalConsensus(network)),
-				DescendOrigin(X1(origin_location)),
-				WithdrawAsset(buy_execution_fee.clone().into()),
-				BuyExecution { fees: buy_execution_fee.clone(), weight_limit: Unlimited },
-				SetAppendix(
-					vec![
-						RefundSurplus,
-						DepositAsset {
-							assets: buy_execution_fee.into(),
-							beneficiary: (
-								Parent,
-								Parent,
-								GlobalConsensus(network),
-								origin_location,
-							).into(),
-						},
-					]
-					.into(),
-				),
-			]
-		};
+impl<G: Get<[u8; 2]>, F: XcmFeeGetter> sp_runtime::traits::TryConvert<VersionedMessage, Xcm<()>>
```

Comment:
Please use a type alias for `[u8; 2]`, to increase descriptiveness.

```suggestion
impl<G: Get<CallIndex>, F: XcmFeeGetter> sp_runtime::traits::TryConvert<VersionedMessage, Xcm<()>>
```